File: healthy/solutions/martinvl.java
/*
@EXPECTED_RESULTS@: CORRECT
*/

import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Scanner;

public class martinvl {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int n = in.nextInt();
        int d = in.nextInt();

        PriorityQueue<Integer> q = new PriorityQueue<Integer>();

        for (int i = 0; i < n; ++i) {
            q.add(in.nextInt());
        }

        PriorityQueue<int[]> q2 = new PriorityQueue<int[]>(d+1, new Comparator<int[]>() {
            public int compare(int[] lhs, int[] rhs) {
                return lhs[0] - rhs[0];
            }
        });

        for (int i = 0; i < d; ++i) {
            q2.add(new int[]{in.nextInt(), in.nextInt()});
        }

        int numMinisters = 0;

        for (int day = 0; !q.isEmpty() || !q2.isEmpty(); ++day) {
            if (q.isEmpty()) {
                day = q2.peek()[0];
            }

            while (!q2.isEmpty() && day >= q2.peek()[0]) {
                int[] p = q2.poll();
                q.add(p[0] + p[1]);
            }

            boolean resigned = false;

            if (q.poll()-1 < day) {
                resigned = true;
            }

            while (!q.isEmpty() && q.peek()-1 <= day) {
                resigned = true;
                q.poll();
            }

            if (resigned) {
                ++numMinisters;
            }
        }

        System.out.println(numMinisters);
    }
}

File: healthy/solutions/martinvl.py
# @EXPECTED_RESULTS@: CORRECT

from sys import stdin
from heapq import heapify, heappush, heappop

n, d = map(int, stdin.next().split())
q, q2 = [], []

if n:
    q = map(int, stdin.next().split())
    heapify(q)

if d:
    q2 = [map(int, line.split()) for line in stdin]
    heapify(q2)

num_ministers = 0
day = 0

while q or q2:
    if not q:
        day = q2[0][0]

    while q2 and day >= q2[0][0]:
        heappush(q, sum(heappop(q2)))
        continue

    resigned = False

    if heappop(q)-1 < day:
        resigned = True

    while q and q[0]-1 <= day:
        resigned = True
        heappop(q)

    if resigned:
        num_ministers += 1

    day += 1

print num_ministers
